### ISSUE TITLE
Stack of layers `nn.Sequential` does not respect Dropout layer

### ISSUE THREAD

**Issue by daskol** [google/flax#3017]
What is the designed way to use `nn.Dropout` as a part of `nn.Sequential`? It turns out that `kwargs` are passed only to the first layer and this is the issue since it supposes to use `training` keyword argumeng of `__call__` to enable or disable dropout layer 
 in `flax` tutorials. So, `training` in the code below does not affect dropout mode.

```python
block = nn.Sequential([
    nn.LayerNorm(),
    nn.SelfAttention(self.embedding_size),
    nn.Dropout(self.rate),
])

output = block(input, determenistic=not training)   # No affect on nn.Dropout.
```

From my perspective, `nn.Sequential` has design issues and it is not clear actually how to forward runtime arguments like `training` or `determenistic`.

**Comment by canyon289** [comment]
I have this same question. Is there an example of how to use dropout when not using `nn.compact` but instead using the `setup` pattern

![image](https://user-images.githubusercontent.com/7213793/232327532-caab1390-f5fd-48f4-8903-825543d2b852.png)

**Comment by andsteing** [comment]
Yes indeed, the `nn.Sequential` combinator only applies the `*args` and `**kwargs` to the first child module:

https://github.com/google/flax/blob/76dec35f09bc0282c435627045404584fc1c19f3/flax/linen/combinators.py#L73-L85

We support the pattern of returning a dictionary/tuple to pass args from one layer to the next.

So in your case I would probably be better to use a sequence of layers and **implement that loop in your code**.

(Updating the existing code to forward `*args` and `**kwargs` is not backwards compatible – with defaultargs – so I don't think we would want to make that change).

**Comment by cgarciae** [comment]
Currently its not easy to specify arguments for submodules defined inside `Sequential` , this issue was discussed in #2131. In theory `flags` could solve this by allowing users to specify some top-level values e.g (following code does NOT work):

```python
module = nn.Sequential([
    nn.LayerNorm(),
    nn.SelfAttention(embedding_size),
    nn.Dropout(rate),
])
...
y = module.apply(..., flags={'deterministic': False})
```
Currently flags are implemented for `Scope` but they are not exposed, they are only used to implement `is_initializing`. Maybe we should revisit the issue of exposing flags as its been also been brought up recently in #2928.

**Comment by canyon289** [comment]
Thank you for this answer!
